source/driver_run.cpp

```diff
@@ -2,86 +2,134 @@
     {
-        ModuleBase::WARNING_QUIT("driver","to use LCAO basis, compile with __LCAO");
+        ModuleBase::WARNING_QUIT("driver", "to use LCAO basis, compile with __LCAO");
     }
 #endif
 
     // the life of ucell should begin here, mohan 2024-05-12
     // delete ucell as a GlobalC in near future
     GlobalC::ucell.setup_cell(GlobalV::stru_file, GlobalV::ofs_running);
 
-    //! 3: initialize Esolver and fill json-structure 
-    p_esolver->before_all_runners(INPUT, GlobalC::ucell);
+    //! 3: initialize Esolver and fill json-structure
+#ifdef __LCAO
+    if (GlobalV::ESOLVER_TYPE == "lr")
+        // use constructor rather than Init function to initialize reference (instead of pointers) to ucell
+        if (GlobalV::GAMMA_ONLY_LOCAL)
+            p_esolver = new ModuleESolver::ESolver_LRTD<double, double>(INPUT, GlobalC::ucell);
+        else if (GlobalV::NSPIN < 4)
+            p_esolver = new ModuleESolver::ESolver_LRTD<std::complex<double>, double>(INPUT, GlobalC::ucell);
+        else
+            throw std::runtime_error("LR-TDDFT is not implemented for spin polarized case");
+    else
+#endif
+        p_esolver->before_all_runners(INPUT, GlobalC::ucell);
 
-    // this Json part should be moved to before_all_runners, mohan 2024-05-12
+        // this Json part should be moved to before_all_runners, mohan 2024-05-12
 #ifdef __RAPIDJSON
     Json::gen_stru_wrapper(&GlobalC::ucell);
 #endif
 
     const std::string cal_type = GlobalV::CALCULATION;
 
-    //! 4: different types of calculations 
-    if(cal_type == "md")
+    //! 4: different types of calculations
+    if (cal_type == "md")
     {
         Run_MD::md_line(GlobalC::ucell, p_esolver, INPUT.mdp);
     }
-    else if(cal_type == "scf" 
-         || cal_type == "relax" 
-         || cal_type == "cell-relax")
+    else if (cal_type == "scf" || cal_type == "relax" || cal_type == "cell-relax")
     {
         Relax_Driver rl_driver;
         rl_driver.relax_driver(p_esolver);
     }
     else
     {
         //! supported "other" functions:
-        //! nscf(PW,LCAO), 
-        //! get_pchg(LCAO), 
-        //! test_memory(PW,LCAO), 
+        //! nscf(PW,LCAO),
+        //! get_pchg(LCAO),
+        //! test_memory(PW,LCAO),
         //! test_neighbour(LCAO),
-        //! get_S(LCAO), 
+        //! get_S(LCAO),
         //! gen_bessel(PW), et al.
         const int istep = 0;
         p_esolver->others(istep);
     }
 
-
     //! 5: clean up esolver
     p_esolver->after_all_runners();
-    ModuleESolver::clean_esolver(p_esolver);
 
+#ifdef __LCAO
+    //---------beyond DFT: set up the next ESolver---------
+    if (INPUT.esolver_type == "ks-lr")
+    {
+        std::cout << "setting up the esolver for excited state" << std::endl;
+        // ModuleESolver::ESolver_KS_LCAO* p_esolver_lcao_tmp = dynamic_cast<ModuleESolver::ESolver_KS_LCAO<double,
+        // double>*>(p_esolver);
+        ModuleESolver::ESolver* p_esolver_lr = nullptr;
+        if (INPUT.gamma_only)
```

Comment:
move this part to functions like init_esolver

Reply:
This part (initializing the 2nd esolver) is called **after** `ucell.setup_cell` and all the functions except for `clean_esolver` of the 1st esolver, but `init_esolver` is called **before** them. 
If `init_esolver` can be moved after `ucell.setup_cell`, maybe it's possible to copy the 1st-esolver-functions in the `if("ks-lr")` branch and share the 2nd-esolver-functions here with other branches. -- What do you think of changing like this?

Reply:
I've changed it according to the above statement in the newest commit.